### PR TITLE
Tokenize ShimmerView

### PR DESCRIPTION
**Pull request by jeaniehuynh**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

Tokenization of ShimmerView. Note that this change changes APIs and is thus a breaking change.

- ios/FluentUI.Demo/FluentUI.Demo/Demos.swift
  - Move ShimmerView demo controller to vNext section
- ios/FluentUI.Demo/FluentUI.Demo/Demos/ShimmerLinesViewDemoController.swift
  - Add new parameters to ShimmerView and ShimmerLinesView inits
- ios/FluentUI.xcodeproj/project.pbxproj
  - Add ShimmerTokens file
- ios/FluentUI/Shimmer/ShimmerLinesView.swift
  - Add lineCount, firstLineFillPercent, and lastLineFillPercent to ShimmerLines init
  - Update to use tokens where available
- ios/FluentUI/Shimmer/ShimmerTokens.swift
  - Add tokens for Shimmer and ShimmerLines
- ios/FluentUI/Shimmer/ShimmerView.swift
  - Move ShimmerStyle to ShimmerTokens
  - Add shimmerStyle, shimmersLeafViews, and usesTextHeightForLabels to Shimmer init
  - Update to use tokens where available

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| https://user-images.githubusercontent.com/31874971/173157760-a5348f0f-c079-4e50-9758-c4b87f79100f.mp4 | https://user-images.githubusercontent.com/31874971/173157570-7ac11d4e-d972-48ab-b7df-1b2de91ee3be.mp4 |
### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [x] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/1004)

### CODE REVIEW

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/ShimmerLinesViewDemoController.swift [hunk 1]
```diff
@@ -33,21 +33,25 @@ class ShimmerViewDemoController: DemoController {
...
-            shimmerView.shimmersLeafViews = shimmersLeafViews
-            shimmerView.usesTextHeightForLabels = true
-            shimmerView.labelHeight = -1 // Must be < 0 so we actually use the bool usesTextHeightForLabels
             containerView.addSubview(shimmerView)
             return containerView
         }
 
-        let shimmeringImageView = { (shimmerStyle: ShimmerStyle) -> UIView in
-            let imageView = UIImageView(image: UIImage(named: "PlaceholderImage")?.withTintColor(Colors.Shimmer.tint, renderingMode: .alwaysOriginal))
+        let tintColor = DynamicColor(light: ColorValue(0xF1F1F1) /* gray50 */,
```
Comment (harrieshin) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/ShimmerLinesViewDemoController.swift:46]: why are we not directly referencing fluent colors and rather using hex values?

Comment (jeaniehuynh) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/ShimmerLinesViewDemoController.swift:46]: Oops! Got in the habit of using DynamicColor for the tokens, will change back to UIColor in next iteration

**Comment by harrieshin** [comment]
lets add a note this is a breaking api change

File: ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift [hunk 2]
```diff
@@ -0,0 +1,104 @@
+//
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import UIKit
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift:6]: something is weird. I would have expected git to pick up this was a move from the "old location"

Comment (harrieshin) [reply ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift:6]: wait... what's the standnard here? why are we moving controls to vnext folder? i thought we only did that if we need side by side?

File: ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            return tint
+        }
+    }
+
+    open var darkGradient: ColorValue {
+        return globalTokens.neutralColors[.black]
+    }
+
+    open var tint: DynamicColor {
+        return DynamicColor(light: ColorValue(0xF1F1F1) /* gray50 */,
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift:53]: this should be updated?

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/ShimmerLinesViewDemoController.swift [hunk 4]
```diff
@@ -33,21 +33,25 @@ class ShimmerViewDemoController: DemoController {
...
-            shimmerView.shimmersLeafViews = shimmersLeafViews
-            shimmerView.usesTextHeightForLabels = true
-            shimmerView.labelHeight = -1 // Must be < 0 so we actually use the bool usesTextHeightForLabels
             containerView.addSubview(shimmerView)
             return containerView
         }
 
-        let shimmeringImageView = { (shimmerStyle: ShimmerStyle) -> UIView in
-            let imageView = UIImageView(image: UIImage(named: "PlaceholderImage")?.withTintColor(Colors.Shimmer.tint, renderingMode: .alwaysOriginal))
+        let tintColor = UIColor(light: Colors.gray50,
```
Comment (harrieshin) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/ShimmerLinesViewDemoController.swift:46]: hard to say. is this tint color actually different from the default fluent style?

File: ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+@objc(MSFShimmerLinesView)
+open class ShimmerLinesView: ShimmerView {
+
+    /// Creates the ShimmerLinesView
+    /// - Parameters:
+    ///   - lineCount: Number of lines that will shimmer in this view. Use 0 if the number of lines should fill the available space.
+    ///   - firstLineFillPercent: The percent the first line (if 2+ lines) should fill the available horizontal space
+    ///   - lastLineFillPercent: The percent the last line should fill the available horizontal space.
+    @objc public init(lineCount: Int,
+                      firstLineFillPercent: CGFloat,
```
Comment (harrieshin) [inline ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift:20]: before , all of these param had default values.

Comment (harrieshin) [reply ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift:20]: also is this meant to be convinience init method?

File: ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift (hunk 2 above)
Comment (jeaniehuynh) [reply ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift:6]: Hm yeah, in GitHub Desktop, it showed up as a move.

It's my understanding that tokenized controls get moved to the vNext folder once they're tokenized (Button is the only control in the vNext folder that I see with a legacy file/folder maintained)

Comment (jeaniehuynh) [reply ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift:6]: I just looked at the first commit and it shows up as a move:
image

File: ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                            lightHighContrast: ColorValue(0x919191) /* gray400 */,
+                            dark: aliasTokens.backgroundColors[.surfaceQuaternary].dark,
+                            darkHighContrast: ColorValue(0x919191) /* gray400 */)
+    }
+
+    /// the width of the gradient in the animation
```
[the remark this answers is not in the document]
Comment (sophialee0416) [reply ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift:59]: (and line 80)

File: ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        switch style {
+        case .concealing:
+            return DynamicColor(light: globalTokens.neutralColors[.white],
+                                dark: globalTokens.neutralColors[.grey8])
+        case .revealing:
+            return tint
+        }
+    }
+
+    open var darkGradient: ColorValue {
```
Comment (sophialee0416) [inline ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift:48]: Documentation here and at 49?

File: ios/FluentUI/Vnext/Shimmer/ShimmerView.swift [hunk 8]
```diff
@@ -5,125 +5,9 @@
...
-        }
-    }
-
-    /// Determines whether the shimmer is a revealing shimmer or a concealing shimmer
-    @objc open var shimmerStyle: ShimmerStyle {
-        didSet {
-            setNeedsLayout()
-        }
-    }
+open class ShimmerView: UIView, TokenizedControlInternal, ControlConfiguration {
```
Comment (sophialee0416) [inline ios/FluentUI/Vnext/Shimmer/ShimmerView.swift:10]: Shouldn't TokenizedControlInternal be enough? I thought ControlConfiguration was for our SwiftUI controls

File: ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift (hunk 3 above)
Comment (jeaniehuynh) [reply ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift:53]: I thought to use DynamicColor here so I could use the aliasTokens for at least that one color, thoughts?

File: MicrosoftFluentUI.podspec [hunk 9]
```diff
@@ -281,7 +281,7 @@ fi', :execution_position => :before_compile }
     shimmer_ios.platform = :ios
     shimmer_ios.dependency 'MicrosoftFluentUI/Core_ios'
     shimmer_ios.dependency 'MicrosoftFluentUI/Utilities_ios'
-    shimmer_ios.source_files = ["ios/FluentUI/Shimmer/**/*.{swift,h}"]
+    shimmer_ios.source_files = ["ios/FluentUI/Vnext/Shimmer/**/*.{swift,h}"]
```
Comment (mischreiber) [inline MicrosoftFluentUI.podspec:284]: @rdeassis is this still something we want to do? What's the value of having these controls live in a different folder, especially if there's only one version of the control in the project at a time?

File: ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift (hunk 3 above)
Comment (mischreiber) [reply ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift:53]: I think this code is right? @harrieshin what kind of update were you thinking?

File: ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    open var cornerRadius: CGFloat { globalTokens.borderRadius[.medium] }
+
+    /// Corner radius on each UILabel. Set to  0 to disable and use default `cornerRadius`.
+    open var labelCornerRadius: CGFloat { globalTokens.borderRadius[.small] }
+
+    /// Height of shimmering labels.
+    open var labelHeight: CGFloat { 11.0 }
+
+    /// Spacing between (if lines > 1)
+    open var labelSpacing: CGFloat { 12.0 }
```
Comment (mischreiber) [inline ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift:81]: `aliasTokens.spacing[.small]`

File: ios/FluentUI/Vnext/Shimmer/ShimmerView.swift (hunk 8 above)
Comment (mischreiber) [reply ios/FluentUI/Vnext/Shimmer/ShimmerView.swift:10]: Correct, `ControlConfiguration` should not be used here.

File: ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+                      lastLineFillPercent: CGFloat) {
+        self.lineCount = lineCount
+        self.firstLineFillPercent = firstLineFillPercent
+        self.lastLineFillPercent = lastLineFillPercent
+
+        super.init()
+    }
+
+    required public init?(coder: NSCoder) {
```
Comment (mischreiber) [inline ios/FluentUI/Vnext/Shimmer/ShimmerLinesView.swift:29]: Can you include a `@available(*, unavailable)` before this line as well?

File: ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift (hunk 3 above)
Comment (harrieshin) [reply ios/FluentUI/Vnext/Shimmer/ShimmerTokens.swift]: why are we using hex values at all? shouldn't we always be using the gray colors directly?

**Review by rdeassis — COMMENTED** [review]
@jeaniehuynh have you found out any potential reasons why the checks did not start running?
It seems like you created this branch from main instead of  fluent2-tokens? (Not sure if this would have any impact on the checks)
I've also quickly looked around and it seems that closing and reopening the PR did the trick for some.

**Comment by jeaniehuynh** [comment]
> @jeaniehuynh have you found out any potential reasons why the checks did not start running? It seems like you created this branch from main instead of fluent2-tokens? (Not sure if this would have any impact on the checks) I've also quickly looked around and it seems that closing and reopening the PR did the trick for some.

@rdeassis I'm not sure why! I thought I saw a thread about it happening before, but can't remember where I saw it nor find it now. I just tried closing and reopening the PR! Also, this branch should be off of fluent2-tokens (I see the vNext folder in this branch which doesn't exist in main).

**Comment by jeaniehuynh** [comment]
^ Just tried pushing a blank commit to restart checks

**Pull request merged**